packages/agent/src/oidc-v2.ts

```diff
@@ -0,0 +1,689 @@
+  /** PermissionGrants that are to be sent to the provider */
+  permission_requests: ConnectPermissionRequests;
+};
+
+// old, ignore, do not use. here for comparison.
+// export interface AuthorizationRequestObject {
+//   claims?: { id_token: { delegation_grants: any; [key: string]: any } };
+//   client_id: string;
+//   client_metadata?: { client_uri?: string; [key: string]: any };
+//   code_verifier?: string;
+//   nonce: string;
+//   redirect_uri: string;
+//   state?: string;
+//   [key: string]: any;
+// }
+
+/** An auth response that is compatible with both Web5 Connect and (hopefully, WIP) OIDC SIOPv2 */
+export type HybridAuthResponse = SIOPv2AuthResponse & Web5ConnectAuthResponse;
+
+/** The fields for an OIDC SIOPv2 Auth Repsonse */
+export type SIOPv2AuthResponse = {
+  /** Issuer Identifier for the Issuer of the response. */
+  iss: 'https://self-issued.me/v2';
+  /** Subject Identifier. A locally unique and never reassigned identifier
+   * within the Issuer for the End-User, which is intended to be consumed
+   * by the Client. */
+  sub: string;
+  /** Audience(s) that this ID Token is intended for. It MUST contain the
+   * OAuth 2.0 client_id of the Relying Party as an audience value. */
+  aud: string;
+  /** Time at which the JWT was issued. */
+  iat: number;
+  /** Expiration time on or after which the ID Token MUST NOT be accepted
+   * for processing. */
+  exp: number;
+  /** Time when the End-User authentication occurred. */
+  auth_time?: number;
+  /** b64url encoded nonce used to associate a Client session with an ID Token, and to
+   * mitigate replay attacks. */
+  nonce?: string;
+  /** Custom claims. */
+  [key: string]: any;
+};
+
+/** The fields for an Web5 Connect Auth Response */
+export type Web5ConnectAuthResponse = {
+  delegation_grants: any[];
+  privateKeyJwks: Jwk[];
+};
+
+/** Represents the different OIDC endpoint types.
+ * 1. `pushedAuthorizationRequest`: client sends {@link PushedAuthRequest} receives {@link PushedAuthResponse}
+ * 2. `authorize`: provider gets the {@link HybridAuthRequest} JWT that was stored by the PAR
+ * 3. `callback`: provider sends {@link HybridAuthResponse} to this endpoint
+ * 4. `token`: client gets {@link HybridAuthResponse} from this endpoint
+ */
+type OidcEndpoint =
+  | 'pushedAuthorizationRequest'
+  | 'authorize'
+  | 'callback'
+  | 'token';
+
+/**
+ * Gets the correct OIDC endpoint out of the {@link OidcEndpoint} options provided.
+ * Handles a trailing slash on baseURL
+ *
+ * @param {Object} options the options object
+ * @param {string} options.baseURL for example `http://foo.com/connect/
+ * @param {OidcEndpoint} options.endpoint the OIDC endpoint desired
+ * @param {string} options.authParam this is the unique id which must be provided when getting the `authorize` endpoint
+ * @param {string} options.tokenParam this is the random state as b64url which must be provided with the `token` endpoint
+ */
+function buildOidcUrl({
```

Comment:
Design pattern: not seeing the benefit of "data coupling" here given that half of the params are contextual.

Would prefer to have 4 independent methods with no coupling. Further still, most of the URL can just be constructed in-line within the caller, since no one is reusing the specific URL constructions.

Reply:
As someone who got to use this method in multiple places it was so helpful to me to centralize the routes and understand the order they're used it and the params that belong to each route. Inlining would never do that. Splitting into 4 is just a mess. The coupling is the point, it's helpful for understanding.

Reply:
> understand the order they're used it and the params that belong to each route

Got it, I see your intent now why you documented 1. ~ 4. steps here inline. The documentation is awesome and super useful for sure, but it feels misplaced for two reasons:
1. Describing input output of each URL for a function called `buildOidcUrl` feels out of place, reader normally wouldn't be expecting to see param info decribed here.
1. This info would probably be more discoverable and useful at the top of the entry point of (`connect-v2.ts`).

 I disagree quite strongly on "Splitting into 4 is just a mess" statement.

It is undeniably useful for the developer (you) of the feature to structure the code this way during development. But for the reader of this function (me), it is not obvious what params can be used/optional/required/in-combination with which endpoint without reading the documentation or code carefully (not self-documenting, quoting from your response in another comment), coupled with the observation that this is not a utility function, lead me to think that the alternative is more desirable than the coincidental coupling in code.

In my view this is an example of "coincidental coupling", which often makes the code look neat on the surface, but require avoidable context to understand.

Apologies for geeking out on this one, honestly there is not a lot of code here to be that confused! But hopefully I explained my principle in thinking. Bottom line is, if you don't feel strongly, please change it, if you feel strongly don't change it, I won't die on this hill either and block the PR on this. I can always change it later if it bothers me enough. 🙏

Reply:
Ok but let me call out: this is the only place outside of the `dwn-server` repo where the routes are centralized, constructed, and then returned.  A comment that does the same thing would be double the code. It would be 4 methods plus a giant comment the size of them. When instead we have a method that accomplishes both tasks. The method is simple to understand and requires no special context. We could add 4 overloads but I had decided it wasn't worth it...

Reply:
> honestly there is not a lot of code here to be that confused

Right that's exactly what I'm saying. It's easily readable to the extent that there's not going to be any confusion as to what it does. No one who knows how to write code is going to be confused by this method, so it shouldn't be a target for some sort of design pattern optimization.

Reply:
My apology for being a little confused when reading how the code works, not much though, just little, hence the suggestion. But seems like you feel pretty strongly about this one, so we'll stop at this.